### PR TITLE
validate role_name for documented criteria

### LINKED ISSUE #75023: Return better error when role name contains invalid characters for collections
### Summary

I have a playbook that looks like this:

```
---
- hosts: host1, host2
  roles:
    - role: namespace.my_collection.common
      tags: common

    - role: namespace.my_collection.autoupdate-apt
      tags: common
```
If I try to use it It reports that the role couldn't be found. If I change name of the autoupdate-apt role to autoupdate_apt in the collection and the playbook it is found and executed

### Issue Type

Bug Report

### Component Name

collections

### Ansible Version

```console
$ ansible --version
ansible [core 2.11.1] 
  config file = /home/leah/example/ansible.cfg
  configured module search path = ['/home/leah/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  ansible collection location = /home/leah/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.9.5 (default, May 24 2021, 12:50:35) [GCC 11.1.0]
  jinja version = 3.0.1
  libyaml = True
```

### Configuration

```console
$ ansible-config dump --only-changed
DEFAULT_MANAGED_STR(/home/leah/example/ansible.cfg) = This file is managed by ansible - please do not edit it as changes will be reverted.
```

### OS / Environment

Current Arch Linux

### Steps to Reproduce

Create a collection with a role containing a - in it's name
Try to use this role according to the playbook snippet above.

### Expected Results

The role with - in it's name is found and executed 

### Actual Results

```console
ERROR! the role 'namespace.my_collection.autoupdate-apt' was not found in /home/leah/example/roles:/home/leah/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles:/home/leah/example/

The error appears to be in '/home/leah/example/site.yml': line 7, column 7, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


    - role: namespace.my_collection.autoupdate-apt
      ^ here
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by Noel-Voelin**
add function to validate role names

##### SUMMARY
I added a little function that is validating the role_name according to the criteria documented in https://docs.ansible.com/ansible/latest/dev_guide/developing_collections_structure.html#roles-directory

This is my first contribution so pelase feel free to come up with plenty of corrections and/or let me know what I have not considered.

Fixes #75023

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
update of RoleDefinition error handling

##### ADDITIONAL INFORMATION
Normally, I would solve this issue using a regular expression in python. However, I did not want to add the re module. If this would have been the way to go, let me know and I apply the changes accordingly.

### CODE REVIEW

**Comment by Noel-Voelin** [comment]
I does seem that the tests are failing because some roles don't actually match the documented conditions. Can I get some insight on how I am supposed to proceed or what I am missing?

**Review by bcoca — CHANGES_REQUESTED** [review]
most existing roles are OUTSIDE of collections, so validating them all as if they were part of a collection is incorrect

This should be a warning in the collection finder code that searches for the role as the non collection search does accept a wider range of characters in the names.

File: lib/ansible/playbook/role/definition.py [hunk 1]
```diff
@@ -228,6 +231,22 @@ def _split_role_params(self, ds):
...
+        https://docs.ansible.com/ansible/latest/dev_guide/developing_collections_structure.html#roles-directory and
+        raise a distinct AnsibleError in case of violation
+        '''
+
+        parts = role_name.split(".")
+        for part in parts:
+            # make sure each part starts alphanumerical
+            # remove legal special character before checking if it's alphanumerical
+            # make sure each part is fully lowercase
+            if not part.startswith('_') and not part.replace('_', '').isalnum() and part.islower():
```
Comment (s-hertel) [inline lib/ansible/playbook/role/definition.py]: This [naming restriction](https://docs.ansible.com/ansible/latest/dev_guide/developing_collections_structure.html#roles-directory) is specific to importing the role into Galaxy. The requirement just to use a role in a collection is that it needs to be a valid python identifier. That's a requirement for collections too - here's how we validate that: https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/collection_loader/_collection_finder.py#L875.

This fix is missing validation for role names provides as strings, e.g.
```
roles:
  - ns.coll.rolename
```

and as already said, validation should only occur for roles in collections.

Comment (Noel-Voelin) [reply lib/ansible/playbook/role/definition.py]: Thank you! I kind of thought that this is handled somewhere already. I will read into it and update my PR soon. Thanks!

**Comment by Noel-Voelin** [comment]
I will reopen this once I have it working properly. No need to keep you in the loop til then. Thanks!

**Pull request closed without merging**